### PR TITLE
Add NavigationAgent Path Debug Visualization

### PR DESCRIPTION
**Pull request by smix8**
Adds path debug visuals for NavigationAgent2D, NavigationAgent3D and NavigationServer.

The NavigationAgent build-in path debug visuals are toggleable per agent in the inspector with `debug_enabled` property. Custom properties like path color or rasterized point size can be also used per agent instead of the global values when `debug_use_custom` is enabled.

agent_debug

As with other navigation debugs the global values can be controlled in the ProjectSettings from ProjectSettings->Debug->Shapes.

agent_debug_ps

While 2D always had an easy time just adding the navigation path points array to a Line2D node for display such an option did not exist for 3D and creating a debug mesh for paths was actually a ton of work and troublesome for beginners. Even for 2D using a node for display was troublesome as it messed with the SceneTree and child functions.

The debug visuals now use the RenderingServer directly, so no unexpected debug Nodes that appear in the SceneTree or RemoteSceneTree and mess up child functions and also no Node overhead. The rendering is updated only once when the agent path or properties change and not every frame to keep runtime performance cost low.

### CODE REVIEW

**Comment by Scony** [comment]
I'll be testing & reviewing this within the next 16h

File: doc/classes/ProjectSettings.xml [hunk 1]
```diff
@@ -519,9 +519,21 @@
...
+			Rasterized size (pixel) used to render navigation agent path points, visible when "Visible Navigation" is enabled in the Debug menu and when an agent has debug enabled.
+		</member>
 		<member name="debug/shapes/navigation/edge_connection_color" type="Color" setter="" getter="" default="Color(1, 0, 1, 1)">
 			Color to display edge connections between navigation regions, visible when "Visible Navigation" is enabled in the Debug menu.
 		</member>
+		<member name="debug/shapes/navigation/enable_agent_paths" type="bool" setter="" getter="" default="true">
+			If enabled, displays navigation agent paths when "Visible Navigation" is enabled in the Debug menu and when an agent has debug enabled.
+		</member>
+		<member name="debug/shapes/navigation/enable_agent_paths_xray" type="bool" setter="" getter="" default="true">
+			If enabled, displays navigation agent paths through geometry when "Visible Navigation" is enabled in the Debug menu and when an agent has debug enabled.
```
Comment (Scony) [inline doc/classes/ProjectSettings.xml:535]: debug paths are visible even if `Visible Navigation` is disabled - IMO that's better as `Visible Navigation` has just a single responsibility. Nevertheless, the descriptions here need to be aligned

File: scene/3d/navigation_agent_3d.cpp [hunk 2]
```diff
@@ -564,3 +617,128 @@ void NavigationAgent3D::_check_distance_to_target() {
...
+	Ref<StandardMaterial3D> debug_agent_path_line_material = NavigationServer3D::get_singleton()->get_debug_navigation_agent_path_line_material();
+	if (debug_use_custom) {
+		if (!debug_agent_path_line_custom_material.is_valid()) {
+			debug_agent_path_line_custom_material = debug_agent_path_line_material->duplicate();
+		}
+		debug_agent_path_line_custom_material->set_albedo(debug_path_custom_color);
+		debug_path_mesh->surface_set_material(0, debug_agent_path_line_custom_material);
+	} else {
+		debug_path_mesh->surface_set_material(0, debug_agent_path_line_material);
+	}
```
Comment (Scony) [inline scene/3d/navigation_agent_3d.cpp:712]: this function is pretty long, consider extracting this part

File: scene/3d/navigation_agent_3d.cpp [hunk 3]
```diff
@@ -564,3 +617,128 @@ void NavigationAgent3D::_check_distance_to_target() {
...
+	if (debug_use_custom) {
+		if (!debug_agent_path_point_custom_material.is_valid()) {
+			debug_agent_path_point_custom_material = debug_agent_path_point_material->duplicate();
+		}
+		debug_agent_path_point_custom_material->set_albedo(debug_path_custom_color);
+		debug_agent_path_point_custom_material->set_point_size(debug_path_custom_point_size);
+		debug_path_mesh->surface_set_material(1, debug_agent_path_point_custom_material);
+	} else {
+		debug_path_mesh->surface_set_material(1, debug_agent_path_point_material);
+	}
```
Comment (Scony) [inline scene/3d/navigation_agent_3d.cpp:736]: consider extracting this part as well

File: scene/2d/navigation_agent_2d.cpp [hunk 4]
```diff
@@ -547,3 +601,109 @@ void NavigationAgent2D::_check_distance_to_target() {
...
+		if (debug_use_custom) {
+			point_size = debug_path_custom_point_size;
+			half_point_size = debug_path_custom_point_size * 0.5;
+		}
+
+		for (int i = 0; i < navigation_path.size(); i++) {
+			const Vector2 &vert = navigation_path[i];
+			Rect2 path_point_rect = Rect2(vert.x - half_point_size, vert.y - half_point_size, point_size, point_size);
+			RenderingServer::get_singleton()->canvas_item_add_rect(debug_path_instance, path_point_rect, debug_path_color);
+		}
```
Comment (Scony) [inline scene/2d/navigation_agent_2d.cpp:706]: this part may be extracted to separate path management from path drawing

File: doc/classes/ProjectSettings.xml (hunk 1 above)
Comment (smix8) [reply doc/classes/ProjectSettings.xml:538]: The Visible Navigation is toggling the NavigationServer3D debug_enabled bool. The bug was that this was not checked.

File: doc/classes/NavigationAgent2D.xml [hunk 5]
```diff
@@ -111,6 +111,21 @@
 		<member name="avoidance_enabled" type="bool" setter="set_avoidance_enabled" getter="get_avoidance_enabled" default="false">
 			If [code]true[/code] the agent is registered for an RVO avoidance callback on the [NavigationServer2D]. When [method NavigationAgent2D.set_velocity] is used and the processing is completed a [code]safe_velocity[/code] Vector2 is received with a signal connection to [signal velocity_computed]. Avoidance processing with many registered agents has a significant performance cost and should only be enabled on agents that currently require it.
 		</member>
+		<member name="debug_enabled" type="bool" setter="set_debug_enabled" getter="get_debug_enabled" default="false">
+			If [code]true[/code] shows debug visuals for this agent.
```
Comment (Scony) [inline doc/classes/NavigationAgent2D.xml:115]: As mentioned before - I don't like the fact that agent debug requires global `Visible Navigation` option to be ON to work, but anyway - in that case IMO we should mention that requirement here - something like _if `true` and `Debug/Visible Navigation` is checked, debug visuals will be drawn for this agent_

File: doc/classes/NavigationAgent3D.xml [hunk 6]
```diff
@@ -114,6 +114,18 @@
 		<member name="avoidance_enabled" type="bool" setter="set_avoidance_enabled" getter="get_avoidance_enabled" default="false">
 			If [code]true[/code] the agent is registered for an RVO avoidance callback on the [NavigationServer3D]. When [method NavigationAgent3D.set_velocity] is used and the processing is completed a [code]safe_velocity[/code] Vector3 is received with a signal connection to [signal velocity_computed]. Avoidance processing with many registered agents has a significant performance cost and should only be enabled on agents that currently require it.
 		</member>
+		<member name="debug_enabled" type="bool" setter="set_debug_enabled" getter="get_debug_enabled" default="false">
+			If [code]true[/code] shows debug visuals for this agent.
```
Comment (Scony) [inline doc/classes/NavigationAgent3D.xml:118]: same here

File: doc/classes/NavigationAgent2D.xml (hunk 5 above)
Comment (smix8) [reply doc/classes/NavigationAgent2D.xml:115]: I agree it should not be tied to the Editor->Debug->Visible Navigation that also toggles on the navmesh debug.

That is currently more a GUI band-aid to toggle on the NavigationServer3D `debug_enabled`. Before, this pr did not check for this global debug so if users disabled the global debug the NavigationAgent would still show debug which can also not be the intended behavior.

Can make the NavigationAgent `debug_enabled` toggle on the debug on the NavigationServer when used, this way users can toggle the debug globally after, at least until the debug GUI is improved.

**Comment by smix8** [comment]
I removed the reliance on the Editor->Debug->Visible Nvaigation and the NavigationServer debug_enabled.

**Review by Scony — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**